src/runtime_src/xocl/core/device.cpp

```diff
@@ -729,11 +729,18 @@ get_cu_memidx() const
       for (auto& cu : get_cu_range())
         mask &= cu->get_memidx_intersect();
 
+      auto mems = m_metadata.get_mem_topology();
       // select first common memory bank index if any
       for (size_t idx=0; idx<mask.size(); ++idx) {
         if (mask.test(idx)) {
-          m_cu_memidx = idx;
-          break;
+	  if (m_cu_memidx == -1)
+		m_cu_memidx = idx; // Select first common memory bank
+	  else {
+	    if (mems && idx < (size_t)mems->m_count && strstr((const char*)mems->m_mem_data[idx].m_tag, "MBG")) {
+		m_cu_memidx = idx;  // Select Group if present prior to bank
+		break;
+	    }
+	  }
```

Comment:
I now finally understand the reason or the reversed loop originally.  If we decide to do this, I think you are right that we should just look for the last common memory bank, which automatically will be a group if it exists.   Reversing the loop does indeed make the code simpler.

Please make sure to comment this code as to the reason for the reverse loop, especially the implicit assumption that groups are last in the group topology section.